libsupport/include/katana/EntityTypeManager.h

```diff
@@ -78,6 +83,181 @@ class EntityTypeManager {
+         ++row) {
+      FieldEntityType field_indices;
+      for (auto bool_property : bool_properties) {
+        if (bool_property.array->IsValid(row) &&
+            bool_property.array->Value(row)) {
+          field_indices.emplace_back(bool_property.field_index);
+        }
+      }
+      for (auto uint8_property : uint8_properties) {
+        if (uint8_property.array->IsValid(row) &&
+            uint8_property.array->Value(row)) {
+          field_indices.emplace_back(uint8_property.field_index);
+        }
+      }
+      if (field_indices.size() > 1) {
+        type_combinations.emplace(field_indices);
+      }
+    }
+
+    // assign a new ID to each unique combination of types
+    for (const FieldEntityType& field_indices : type_combinations) {
+      std::vector<std::string> field_names;
+      for (int i : field_indices) {
+        const std::shared_ptr<arrow::Field>& current_field = schema->field(i);
+        const std::string& field_name = current_field->name();
+        field_names.emplace_back(field_name);
+      }
+      katana::EntityTypeID new_entity_type_id =
+          entity_type_manager->AddNonAtomicEntityType(field_names);
+      type_field_indices_to_id.emplace(
+          std::make_pair(field_indices, new_entity_type_id));
+    }
+
+    // assert that all type IDs (including kUnknownEntityType) and
+    // 1 special type ID (kInvalidEntityType)
+    // can be stored in 8 bits
+    if (entity_type_manager->GetNumEntityTypes() >
+        (std::numeric_limits<katana::EntityTypeID>::max() - size_t{1})) {
+      return KATANA_ERROR(
+          katana::ErrorCode::NotImplemented,
+          "number of unique combination of types is {} "
+          "but only up to {} is supported currently",
+          // exclude kUnknownEntityType
+          entity_type_manager->GetNumEntityTypes() - 1,
+          // exclude kUnknownEntityType and kInvalidEntityType
+          std::numeric_limits<katana::EntityTypeID>::max() - 2);
+    }
+
+    // assign the type ID for each row
+    for (int64_t row = 0; row < num_rows; ++row) {
+      FieldEntityType field_indices;
+      for (auto bool_property : bool_properties) {
+        if (bool_property.array->IsValid(row) &&
+            bool_property.array->Value(row)) {
+          field_indices.emplace_back(bool_property.field_index);
+        }
+      }
+      for (auto uint8_property : uint8_properties) {
+        if (uint8_property.array->IsValid(row) &&
+            uint8_property.array->Value(row)) {
+          field_indices.emplace_back(uint8_property.field_index);
+        }
+      }
+      if (field_indices.empty()) {
+        entity_type_ids->at(row) = katana::kUnknownEntityType;
+      } else {
+        katana::EntityTypeID entity_type_id =
+            type_field_indices_to_id.at(field_indices);
+        entity_type_ids->at(row) = entity_type_id;
+      }
+    }
+
+    return ResultSuccess();
+  }
+
   // TODO(amber): delete this method. It's risky
   void Reset() {
```

Comment:
Please see if you can remove this method while you're at it.

Reply:
Remove the `Reset()` method?

Reply:
yes. I don't think we need it. Any use can replaced with `= EntityTypeManager{}`